pkgs/misc/crust/default.nix

```diff
@@ -0,0 +1,154 @@
+    # builtins.match returns 'Null' if the string matches so we have to then check if it's Null
+    # FIXME-QA(Krey): I hate this, but am not aware of better way to write this
+    pname = if (builtins.isNull ((builtins.match "_defconfig$" defconfig)))
+      then "crust-${lib.strings.removeSuffix "_defconfig" defconfig}"
+      else "crust-${defconfig}";
+
+    ###! The package has upstream definitions of configurations for set devices and thus the defconfig has to be always included, beyond that the options are designed to be as flexible as possible to accompany special requirements of various devices
+    ###!
+    ###! Avoid declaring extra configuration (extraConfig) when possible as those should go to the upstream unless you use special usecase device e.g. modded teres-A64
+
+    defconfig = if defconfig == null then
+      throw "crust config not defined!"
+      else defconfig;
+
+    version = if src == null then
+      defaultVersion
+      else version;
+
+    src = if src == null then
+      defaultSrc
+      else src;
+
+    # Only for global crust patches, per-device paches should be included as extras to avoid adding noise which would complicate debugging
+    patches = [] ++ extraPatches;
+
+    postPatch = "";
+
+    nativeBuildInputs = [
+      flex
+      bison
+    ];
+
+    depsBuildBuild = [ buildPackages.stdenv.cc ];
+
+    makeFlags = [
+      # This is double-quotted to prevent string splitting (SC2086)
+      "CROSS_COMPILE=\"${stdenv.cc.targetPrefix}\""
+    ] ++ extraMakeFlags;
+
+    passAsFile = [ "extraConfig" ];
+
+    ###! Phases
+    configurePhase = ''
+      runHook preConfigure
+
+      echo "Configuring for ${defconfig}"
+        make ${defconfig}
+
+      # The variable 'extraConfigPath' is often parsed as blank resulting in a scenario where `cat` is expecting interactive input followed by line-break to continue, this is to reduce the risk of that happening
+      [ -z "$extraConfigPath" ] || {
+        cat $extraConfigPath >> .config
+      }
+
+      runHook postConfigure
+    '';
+
+    installPhase = ''
+      runHook preInstall
+
+      echo "Creating directory in ${installDir}"
+      mkdir -p "${installDir}"
+
+      # Process 'filesToInstall' from declarations
+      for file in ${lib.concatStringsSep " " filesToInstall}; do
+        cp -v "$file" "${installDir}"
+      done
+
+      runHook postInstall
+    '';
+
+    ###! Hardening
+    ###! * 'bindnow' is ignored by the package
+    ###! * 'pie' is already applied by default to musl and disabled for glibc where in our scenario it seems to be unsupported
+    ###! * 'relro' is ignored by the package, likely due to it breaking dynamic objects. Issue https://github.com/crust-firmware/crust/issues/216 has been submitted to verify
+    ###! Refer to https://nixos.org/manual/nixpkgs/stable/#sec-hardening-in-nixpkgs for details
+    hardeningDisable = [
+      "bindnow"
+      "relro"
+      "pie"
+    ];
+
+    # Default Metadata
```

Comment:
```suggestion
```

Reply:
For nix-inexperienced user/developer this is helpful to understand what the code is doing so i don't want to remove it.